crates/driver/src/domain/competition/auction.rs

```diff
@@ -53,33 +90,114 @@ impl Auction {
             };
             std::cmp::Reverse((
                 class,
+                // TODO No need to expose Tokens anymore, just use the auction
                 // If the orders are of the same kind, then sort by likelihood of fulfillment
                 // based on token prices.
-                order.likelihood(&tokens),
+                order.likelihood(&self.tokens),
             ))
         });
 
-        // TODO Filter out orders based on user balance
+        // Fetch balances of each token for each trader.
+        // Has to be separate closure due to compiler bug.
+        let f = |order: &competition::Order| -> (order::Trader, eth::TokenAddress) {
+            (order.trader(), order.sell.token)
+        };
+        let tokens_by_trader = self.orders.iter().map(f).unique();
+        let mut balances: HashMap<
+            (order::Trader, eth::TokenAddress),
+            Result<eth::TokenAmount, crate::infra::blockchain::Error>,
+        > = join_all(tokens_by_trader.map(|(trader, token)| async move {
+            let balance = eth.balance_of(trader.into(), token).await;
+            ((trader, token), balance)
+        }))
+        .await
+        .into_iter()
+        .collect();
+
+        self.orders.retain(|order| {
+            // TODO: We should use balance fetching that takes interactions into account
+            // from `crates/shared/src/account_balances/simulation.rs` instead of hardcoding
+            // an Ethflow exception. https://github.com/cowprotocol/services/issues/1595
```

Comment:
I believe we already have the pre-interaction aware balance fetcher, right?
In that case it might be less effort to immediately use that instead of first introducing all the ethflow machinery for this exception and later dropping that again.

Reply:
Leaving that unaddressed for now as things are not as straight forward than I thought.